File: src/core/system/load/cache.rb
require 'core/schema/tools/dumpjson'
require 'core/system/utils/find_model'
require 'core/schema/code/factory'
require 'digest/sha1'
require 'enso'

module Cache

	def self.hack_prefix
	  if Enso::System.is_javascript() # HACK TO GET ELECTRON RUNNING!!!
	    "../" 
    else
			""
		end
  end
	
  def self.save_cache(name, model, full=false)
    out = find_json(name)
    res = add_metadata(name, model)
    res['model'] = Dumpjson::to_json(model, full)
    File.open(out, 'w+') do |f| 
      f.write(JSON.pretty_generate(res, allow_nan: true, max_nesting: false))
    end
  end

  # this default param is currently not supported by the Rascal compiler.
  # because it captures an earlier parameter.
  def self.load_cache(model, schema, path = find_json(model))
    factory = Factory::SchemaFactory.new(schema)
    #STDERR.puts("## loading cache for: #{model} (#{path})")
    json = Enso::System.readJSON(path)
    res = Dumpjson::from_json(factory, json['model'])
    res.factory.file_path[0] = path
    json['depends'].each {|dep| res.factory.file_path << dep['filename']}
    res
  end

  def self.check_dep(name)
    begin
      path = find_json(name)
      json = Enso::System.readJSON(path)

      #check that the source file has not changed
      #check that none of the dependencies have changed
      check_file(json) && json['depends'].all? {|e| check_file(e)}
    rescue Errno::ENOENT => e
      false
    end
  end

  def self.clean(name=nil)
    cache_path = "cache/"
    if name.nil? #clean everything
      if File.exists?("#{cache_path}")
        Dir.foreach("#{cache_path}") do |f|
          if f.end_with?(".json")
            File.delete("#{cache_path}#{f}")
          end
        end
        true
      else
        false
      end
    else
      if ['schema.schema', 'schema.grammar', 'grammar.schema', 'grammar.grammar'].include?(name)
        false
      else
        if File.exists?(find_json(name))
          File.delete(find_json(name))
          true
        else
          false
        end
      end
    end
  end

  def self.find_json(name)
    if ['schema.schema', 'schema.grammar', 'grammar.schema', 'grammar.grammar'].include?(name)
      "#{hack_prefix}core/system/boot/#{name}.json"
    else
	    cache_path = "#{hack_prefix}cache/"
      "#{cache_path}#{name}.json"
    end
  end
  
  def self.check_file(element)
    if Digest.nil?
      true
    else
	    path = element['source']
	    checksum = element['checksum']
	    begin
	      readHash(path)==checksum
	    rescue
	      false
	    end
	  end
  end

  def self.get_meta(name)
    e = {filename: name}
    begin
	    FindModel::find_model(name) do |path|
	      e['source'] = path
	      e['date'] = File.ctime(path)
	      e['checksum'] = readHash(path)
	    end
	  rescue
      e['source'] = "SYNTHETIC"
      e['date'] = Time.new
    end
    e
  end

  def self.add_metadata(name, model)
    if name==nil
      e = {filename: 'MetaData'}
    else
      e = get_meta(name)
      type = name.split('.')[-1]

      #vertical deps up the metamodel stack
      deps = []
      deps << get_meta("#{type}.grammar")

      #analyze horizontal dep
      #something to do with imports
      # STDERR.puts "METADATA #{model.factory.file_path}"
      if model.factory.file_path.size > 0
	      model.factory.file_path[1..-1].each {|fn| deps << get_meta(fn.split("/")[-1])}
	    end
      e['depends'] = deps
    end
    e
  end

  def self.readHash(path)
    hashfun = Digest::SHA1.new
    fullfilename = path
    open(fullfilename, "r") do |io|
      while (!io.eof)
        readBuf = io.readpartial(50)
        hashfun.update(readBuf)
      end
    end
    hashfun.to_s
  end

end

File: src/js/core/system/load/cache.js
'use strict'

//// Cache ////

var cwd = process.cwd() + '/';
var Dumpjson = require(cwd + "core/schema/tools/dumpjson.js");
var FindModel = require(cwd + "core/system/utils/find_model.js");
var Factory = require(cwd + "core/schema/code/factory.js");
var Sha1 = require(cwd + "digest/sha1.js");
var Enso = require(cwd + "enso.js");

var Cache;

var hack_prefix = function() {
  var self = this;
  if (Enso.System.is_javascript()) {
    return "../";
  } else {
    return "";
  }
};

var save_cache = function(name, model, full = false) {
  var self = this, out, res;
  out = Cache.find_json(name);
  res = Cache.add_metadata(name, model);
  res .set$("model", Dumpjson.to_json(model, full));
  return Enso.File.open(function(f) {
    return f.write(JSON.pretty_generate(res, Enso.EMap.new({allow_nan: true, max_nesting: false})));
  }, out, "w+");
};

var load_cache = function(model, schema, path = Cache.find_json(model)) {
  var self = this, factory, json, res;
  factory = Factory.SchemaFactory.new(schema);
  json = Enso.System.readJSON(path);
  res = Dumpjson.from_json(factory, json.get$("model"));
  res.factory().file_path() .set$(0, path);
  json.get$("depends").each(function(dep) {
    return res.factory().file_path().push(dep.get$("filename"));
  });
  return res;
};

var check_dep = function(name) {
  var self = this, path, json;
  try {
    path = Cache.find_json(name);
    json = Enso.System.readJSON(path);
    return Cache.check_file(json) && json.get$("depends").all_P(function(e) {
      return Cache.check_file(e);
    });
  } catch (e) {
    return false;
  }
};

var clean = function(name = null) {
  var self = this, cache_path;
  cache_path = "cache/";
  if (name == null) {
    if (Enso.File.exists_P(cache_path)) {
      Dir.foreach(function(f) {
        if (f.end_with_P(".json")) {
          return Enso.File.delete_M(Enso.S(cache_path, f));
        }
      }, cache_path);
      return true;
    } else {
      return false;
    }
  } else if (["schema.schema", "schema.grammar", "grammar.schema", "grammar.grammar"].include_P(name)) {
    return false;
  } else if (Enso.File.exists_P(Cache.find_json(name))) {
    Enso.File.delete_M(Cache.find_json(name));
    return true;
  } else {
    return false;
  }
};

var find_json = function(name) {
  var self = this, cache_path;
  if (["schema.schema", "schema.grammar", "grammar.schema", "grammar.grammar"].include_P(name)) {
    return Enso.S(Cache.hack_prefix(), "core/system/boot/", name, ".json");
  } else {
    cache_path = Enso.S(Cache.hack_prefix(), "cache/");
    return Enso.S(cache_path, name, ".json");
  }
};

var check_file = function(element) {
  var self = this, path, checksum;
  if (Digest == null) {
    return true;
  } else {
    path = element.get$("source");
    checksum = element.get$("checksum");
    try {
      return Cache.readHash(path) == checksum;
    } catch (DUMMY) {
      return false;
    }
  }
};

var get_meta = function(name) {
  var self = this, e;
  e = Enso.EMap.new({filename: name});
  try {
    FindModel.find_model(function(path) {
      e .set$("source", path);
      e .set$("date", Enso.File.ctime(path));
      return e .set$("checksum", Cache.readHash(path));
    }, name);
  } catch (DUMMY) {
    e .set$("source", "SYNTHETIC");
    e .set$("date", Time.new());
  }
  return e;
};

var add_metadata = function(name, model) {
  var self = this, e, type, deps;
  if (name == null) {
    e = Enso.EMap.new({filename: "MetaData"});
  } else {
    e = Cache.get_meta(name);
    type = name.split_M(".").get$(- 1);
    deps = [];
    deps.push(Cache.get_meta(Enso.S(type, ".grammar")));
    if (model.factory().file_path().size_M() > 0) {
      model.factory().file_path().get$(Range.new(1, - 1)).each(function(fn) {
        return deps.push(Cache.get_meta(fn.split_M("/").get$(- 1)));
      });
    }
    e .set$("depends", deps);
  }
  return e;
};

var readHash = function(path) {
  var self = this, hashfun, fullfilename, readBuf;
  hashfun = Digest.SHA1.new();
  fullfilename = path;
  Cache.open(function(io) {
    while (! io.eof()) {
      readBuf = io.readpartial(50);
      hashfun.update(readBuf);
    }
  }, fullfilename, "r");
  return hashfun.to_s();
};

Cache = {
  hack_prefix: hack_prefix,
  save_cache: save_cache,
  load_cache: load_cache,
  check_dep: check_dep,
  clean: clean,
  find_json: find_json,
  check_file: check_file,
  get_meta: get_meta,
  add_metadata: add_metadata,
  readHash: readHash,
};
module.exports = Cache ;
